### PR TITLE
LowPtElectrons: fix for UL FastSim mini v2 and nano v9 workflows

### PR DESCRIPTION
**Pull request by bainbrid**
#### PR description:

- UL mini v2 workflows for FastSim are broken due to an exception thrown by the `LowPtGsfElectronIDProducer` when attempting to access an `edm::Ref` to the missing `generalTracksBeforeMixing` collection in AOD. 
- All details can be found in this issue: https://github.com/cms-sw/cmssw/issues/34774.
- This PR provides an acceptable fix in the RECO chain by suppressing (i.e. removing) the FastSim switch to using tracks from the `generalTracksBeforeMixing` collection in the `GsfElectronCore` and `GsfElectron` objects.
- For re-miniaod (and re-nanoaod) workflows, this PR also provides an acceptable workaround that allows the ID producer to evaluate the BDT model without throwing an exception, while maintaining the expected physics performance. This is achieved by extracting features from the KF track used in the `ElectronSeed` step (obtained from the available `generalTracks` collection in AOD), instead of from the KF track embedded in the `GsfElectronCore` object (which is chosen based on number of shared hits with the `GsfTrack` and obtained from the missing `generalTracksBeforeMixing` collection in AOD). 
- All development was performed in the 10_6_X cycle (see the compare [here](https://github.com/cms-sw/cmssw/compare/CMSSW_10_6_X...bainbrid:LowPtElectrons_ULFastSimFix_106X)), tested with the example recipes provided in the issue [#34774](#34774), and then ported to master (this PR). This work will be back ported to 10_6_X in a future PR.
- Also tested is the behaviour for the nano v9 workflow, which is similarly affected.
- This work was presented to RECO/AT here: https://indico.cern.ch/event/1071803/. The final slide compares physics performance b/w the 10_6_X and 12_1_X cycles (essentially identical).
- Physics and computing performances are ~unaffected. 

#### PR validation:

This PR was validated using cmsDriver commands that performed the GEN->AOD and mini v2 steps, based on those provided [here](https://github.com/cms-sw/cmssw/issues/34774#issue-960576497), for tests with the 10_6_X cycle, and modified versions for tests with the 12_1_X release (this PR). 

@slava77 @jordan-martins @crovelli

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronIDProducer.cc [hunk 1]
```diff
@@ -135,6 +154,29 @@ void LowPtGsfElectronIDProducer::produce(edm::StreamID, edm::Event& event, const
...
+                                              : nullptr;
+            const edm::Ptr<pat::PackedCandidate>* ptr2 =
+                ele.hasUserData("ele2lost") ? ele.template userData<edm::Ptr<pat::PackedCandidate>>("ele2lost")
+                                            : nullptr;
+            if (ptr1 != nullptr && ptr1->isNonnull() && ptr1->isAvailable() && ptr1->get() != nullptr &&
+                ptr1->get()->bestTrack() != nullptr) {
+              return ptr1->get()->bestTrack();
+            } else if (ptr2 != nullptr && ptr2->isNonnull() && ptr2->isAvailable() && ptr2->get() != nullptr &&
+                       ptr2->get()->bestTrack() != nullptr) {
+              return ptr2->get()->bestTrack();
```
Comment (slava77) [inline RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronIDProducer.cc:170]: the string of checks is a bit elaborate and repeated
```suggestion
            bool hasBestTrack = [] (const PackedPtr* ptr) {
              return ptr != nullptr && ptr->isNonnull() && ptr->isAvailable() && ptr->get() != nullptr &&
                ptr->get()->bestTrack() != nullptr
            };
            if (hasBestTrack(ptr1)) {
              return ptr1->get()->bestTrack();
            } else if (hasBestTrack(ptr2)) {
              return ptr2->get()->bestTrack();
```

File: RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc [hunk 2]
```diff
@@ -168,17 +169,28 @@ namespace lowptgsfeleid {
...
+        if (trk.isNonnull()) {
+          eid_trk_p = (float)trk->p();
+          eid_trk_nhits = (float)trk->found();
+          eid_trk_chi2red = (float)trk->normalizedChi2();
+          TVector3 trkTV3(0, 0, 0);
+          trkTV3.SetPtEtaPhi(trk->pt(), trk->eta(), trk->phi());
+          TVector3 eleTV3(0, 0, 0);
+          eleTV3.SetPtEtaPhi(ele.pt(), ele.eta(), ele.phi());
+          trk_dr = eleTV3.DeltaR(trkTV3);
+        }
```
Comment (slava77) [inline RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc:193]:
```suggestion
    if (trk != nullptr || (ele.core().isNonnull() && ele.closestCtfTrackRef().isNonnull()) {
      const reco::Track* tk = trk ? trk : &*ele.closestCtfTrackRef().isNonnull();
      eid_trk_p = tk->p();
      eid_trk_nhits = tk->found();
      eid_trk_chi2red = tk->normalizedChi2();
      TVector3 trkTV3(0, 0, 0);
      trkTV3.SetPtEtaPhi(tk->pt(), tk->eta(), tk->phi());
      TVector3 eleTV3(0, 0, 0);
      eleTV3.SetPtEtaPhi(ele.pt(), ele.eta(), ele.phi());
      trk_dr = reco::deltaR(*tk, ele);
    }
```
- remove repeated code
- remove local `trk` shadowing the input `trk`
- drop `(float)`, which is already clear from lhs
- remove unnecessary and pretty heavy conversion to `TVector3`

File: RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronIDProducer.cc (hunk 1 above)
Comment (bainbrid) [reply RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronIDProducer.cc:170]:
```
plugins/LowPtGsfElectronIDProducer.cc:166:34: error: 'hasBestTrack' cannot be used as a function
  166 |             if (hasBestTrack(ptr1)) {
      |                                  ^
```
@slava77 ?

Comment (slava77) [reply RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronIDProducer.cc:170]: please try `auto hasBestTrack =` and `;` at the end of return.

File: RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronIDProducer.cc [hunk 3]
```diff
@@ -135,6 +154,29 @@ void LowPtGsfElectronIDProducer::produce(edm::StreamID, edm::Event& event, const
...
           return ele.electronID(kUnbiased);
         },
+        [&](auto const& ele) {  // trkFunctor ...
+          if (useGsfToTrack_) {
+            const edm::Ptr<pat::PackedCandidate>* ptr1 =
+                ele.hasUserData("ele2packed") ? ele.template userData<edm::Ptr<pat::PackedCandidate>>("ele2packed")
+                                              : nullptr;
+            const edm::Ptr<pat::PackedCandidate>* ptr2 =
+                ele.hasUserData("ele2lost") ? ele.template userData<edm::Ptr<pat::PackedCandidate>>("ele2lost")
+                                            : nullptr;
```
[the remark this answers is not in the document]
Comment (bainbrid) [reply RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronIDProducer.cc]: Done.

File: RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronIDProducer.cc (hunk 1 above)
Comment (bainbrid) [reply RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronIDProducer.cc]: Done.

File: RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc (hunk 2 above)
Comment (bainbrid) [reply RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc]: Done.

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc [hunk 4]
```diff
@@ -168,18 +169,16 @@ namespace lowptgsfeleid {
...
-      }
+    if (trk != nullptr || (ele.core().isNonnull() && ele.closestCtfTrackRef().isNonnull())) {
+      const reco::Track* tk = trk ? trk : &*ele.closestCtfTrackRef();
+      eid_trk_p = tk->p();
+      eid_trk_nhits = tk->found();
+      eid_trk_chi2red = tk->normalizedChi2();
+      TVector3 trkTV3(0, 0, 0);
+      trkTV3.SetPtEtaPhi(tk->pt(), tk->eta(), tk->phi());
+      TVector3 eleTV3(0, 0, 0);
+      eleTV3.SetPtEtaPhi(ele.pt(), ele.eta(), ele.phi());
```
Comment (slava77) [inline RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc:180]:
```suggestion
```
these temp vectors are not needed anymore

Comment (bainbrid) [reply RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc]: Yep, sorry, removed.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+reconstruction

for https://github.com/cms-sw/cmssw/pull/35181	0212d1d
- code changes are in line with the PR description and the follow up review and the preceding discussion in #34774 
- jenkins tests pass and comparisons with the baseline show differences in `lowPtGsfElectrons.convVtxFitProb` which is computed using `ctfTrack`, which is modified by this PR
- I tested locally that an AODSIM  fastsim file produced in CMSSW_10_6_22 can be processed now with ` --step PAT --procModifiers run2_miniAOD_UL --fast` (in line with the PR description)

**Comment by slava77** [comment]
@bainbrid 
it may be practical to prepare a PR already for 10_6_X.

**Comment by qliphy** [comment]
+1

**Comment by bainbrid** [comment]
> @bainbrid
> it may be practical to prepare a PR already for 10_6_X.

Yes, I already have the branch ready, just need to fold in your suggested changes. Will try to make the PR today.

**Pull request merged**